iceoryx2-bb/system-types/src/file_name.rs

```diff
@@ -61,8 +54,252 @@ semantic_string! {
         }
     }
     false
-  },
-  normalize: |this: &FileName| {
-      *this
-  }
+}
+
+fn invalid_content(value: &[u8]) -> bool {
+    matches!(value, b"" | b"." | b"..")
+}
+
+fn normalize(this: &FileName) -> FileName {
+    *this
+}
+
+semantic_string! {
+  /// Represents a file name. The restriction are choosen in a way that it is platform independent.
+  /// This means characters/strings which would be legal on some platforms are forbidden as well.
+  name: FileName,
+  capacity: FILENAME_LENGTH,
+  invalid_content: invalid_content,
+  invalid_characters: invalid_characters,
+  normalize: normalize
+}
+
+#[derive(Debug, Clone, Copy, Eq)]
+pub struct RestrictedFileName<const CAPACITY: usize> {
+    value: iceoryx2_bb_container::byte_string::FixedSizeByteString<CAPACITY>,
+}
+
+impl<const CAPACITY: usize>
+    iceoryx2_bb_container::semantic_string::internal::SemanticStringAccessor<CAPACITY>
+    for RestrictedFileName<CAPACITY>
+{
+    unsafe fn new_empty() -> Self {
+        static_assert_le::<{ CAPACITY }, { FILENAME_LENGTH }>();
+        static_assert_ge::<{ CAPACITY }, 1>();
+
+        Self {
+            value: iceoryx2_bb_container::byte_string::FixedSizeByteString::new(),
+        }
+    }
+
+    unsafe fn get_mut_string(
+        &mut self,
+    ) -> &mut iceoryx2_bb_container::byte_string::FixedSizeByteString<CAPACITY> {
+        &mut self.value
+    }
+
+    fn is_invalid_content(string: &[u8]) -> bool {
+        if Self::does_contain_invalid_characters(string) {
+            return true;
+        }
+
+        invalid_content(string)
+    }
+
+    fn does_contain_invalid_characters(string: &[u8]) -> bool {
+        if core::str::from_utf8(string).is_err() {
+            return true;
+        }
+
+        invalid_characters(string)
+    }
+}
+
+// BEGIN: serde
+pub(crate) mod visitor_type {
+    pub(crate) struct RestrictedFileName<const CAPACITY: usize>;
+}
+
+impl<'de, const CAPACITY: usize> serde::de::Visitor<'de>
+    for visitor_type::RestrictedFileName<CAPACITY>
+{
+    type Value = RestrictedFileName<CAPACITY>;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a string containing the service name")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        match RestrictedFileName::<CAPACITY>::new(v.as_bytes()) {
+            Ok(v) => Ok(v),
+            Err(v) => Err(E::custom(std::format!(
+                "invalid RestrictedFileName<{}> provided {:?}.",
+                CAPACITY,
+                v
+            ))),
+        }
```

Comment:
This can be simplified
```suggestion
        RestrictedFileName::<CAPACITY>::new(v.as_bytes()).map_err(|v| {
            E::custom(std::format!(
                "invalid RestrictedFileName<{}> provided {:?}.",
                CAPACITY,
                v
            ))
        })
```